include/pagmo/algorithms/gwo.hpp

```diff
@@ -0,0 +1,365 @@
+     */
+
+    population evolve(population pop) const
+    {
+        const auto &prob = pop.get_problem();
+        auto dim = prob.get_nx();
+        auto prob_f_dimension = prob.get_nf();
+        const auto bounds = prob.get_bounds();
+        const auto &lb = bounds.first;
+        const auto &ub = bounds.second;
+        auto NP = pop.size(); // number of agent(wolf) equal to population size
+        auto count = 1u;      // regulates the screen output
+
+        // PREAMBLE-------------------------------------------------------------------------------------------------
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+        if (prob.get_nc() != 0u) {
+            pagmo_throw(std::invalid_argument, "Non linear constraints detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+        if (prob_f_dimension != 1u) {
+            pagmo_throw(std::invalid_argument, "Multiple objectives detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        // Get out if there is nothing to do.
+        if (m_gen == 0u) {
+            return pop;
+        }
+        if (pop.size() < 3u) {
+            pagmo_throw(std::invalid_argument, get_name() + " needs at least 3 individuals in the population, "
+                                                   + std::to_string(pop.size()) + " detected");
+        }
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Some vectors used during evolution are declared.
+        vector_double a_vector(3);          // vector coefficient for encircling prey
+        vector_double c_vector(3);          // vector coefficient for encircling prey
+        vector_double d_vector(3);          // position vector for alpha, beta and delta
+        vector_double x_vector(3);          // vector used to calculate position of an agent in a single dimension
+        double a;                           // coefficient which decrease linearly from 2 to 0 over generations
+        double r1, r2;                      // random coefficient between 0 and 1;
+        std::vector<std::size_t> index_vec; // used to stored sorted index
+        auto agents_position = pop.get_x();
+        auto init_fit = pop.get_f();
+        for (unsigned int i = 0; i != init_fit.size(); ++i)
+            index_vec.push_back(i);
+        std::sort(index_vec.begin(), index_vec.end(),
```

Comment:
This is UB (Undefined Behaviour) whenever the fitness returned by the UDP is NaN or Inf (which is not preventable). In this case the `<` operator is not meeting the compare requirments (https://en.cppreference.com/w/cpp/named_req/Compare)

In Pagmo we thus use https://github.com/esa/pagmo2/blob/master/include/pagmo/detail/custom_comparisons.hpp#L48